cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -255,8 +267,18 @@ private boolean runReadyJobs() throws Exception {
         // may have failed / rolled back.
         Job storedJob = jobQueue.getJob(job.getJobKey());
         if (storedJob == null) {
+          readyJobsIter.remove();
           continue;
         }
+        ProgramScheduleStoreDataset store = Schedulers.getScheduleStore(context, datasetFramework);
+        ProgramScheduleMeta meta = store.getScheduleMeta(job.getSchedule().getScheduleId());
+        if (meta == null || !ProgramScheduleStatus.SCHEDULED.equals(meta.getStatus())
```

Comment:
We can just check whether `job.isToBeDeleted()`. If this is true, that means the schedule was deleted/updated before the state transition from `PENDING_CONSTRAINT` to `PENDING_LAUNCH` is committed. We can just remove the job without launching it. If `job.isToBeDeleted()` is false, that means this job is in `PENDING_LAUNCH` state before the schedule is deleted/updated. We can still launch it.